src/DynamoCore/Visualization/RenderPackageCache.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using Autodesk.DesignScript.Interfaces;
+
+namespace Dynamo.Visualization
+{
+    /// <summary>
+    /// This class controls the collection and distribution of render packages
+    /// </summary>
+    public class RenderPackageCache
+    {
+        private List<IRenderPackage> packages;
+        private Dictionary<Guid, RenderPackageCache> portMap;
+
+        private void AddPort(IRenderPackage package, Guid outputPortId)
+        {
+            if (portMap == null)
+            {
+                portMap = new Dictionary<Guid, RenderPackageCache>();
+            }
+
+            if (!portMap.ContainsKey(outputPortId))
+            {
+                portMap[outputPortId] = new RenderPackageCache();
+            }
+
+            portMap[outputPortId].Add(package);
+        }
+
+        /// <summary>
+        /// Create an empty RenderPackageCache object
+        /// </summary>
+        public RenderPackageCache()
+        {
+            packages = new List<IRenderPackage>();
+            portMap = null;
+        }
+
+        /// <summary>
+        /// Create a RenderPackageCache object containing the given render packages
+        /// </summary>
+        /// <param name="otherPackages">The packages to fill the new cache object.</param>
+        public RenderPackageCache(IEnumerable<IRenderPackage> otherPackages)
+        : this()
+        {
+            packages.AddRange(otherPackages);
+        }
+
+        /// <summary>
+        /// Get the render packages in this cache
+        /// </summary>
+        public IEnumerable<IRenderPackage> Packages
+        {
+            get
+            {
+                return packages;
+            }
+        }
+
+        /// <summary>
+        /// Get the RenderPackageCache object for the given port ID
+        /// </summary>
+        /// <param name="portId">The port ID used to get the sub-cache.</param>
+        public RenderPackageCache GetPortPackages(Guid portId)
+        {
+            if (portMap == null)
+                return null;
+
+            RenderPackageCache portPackages;
+            if (!portMap.TryGetValue(portId, out portPackages))
+                return null;
+
+            if (portPackages == null)
```

Comment:
seems like this is an unnecessary statement, guess useful for debugging though.

Reply:
You are correct, this is totally an unnecessary statement. I will remove it, as it was an unintentional mistake.